artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -512,6 +521,12 @@ private boolean sendMessage(Message message, DeliveryAnnotations deliveryAnnotat
 
       routingContext.setDuplicateDetection(false); // we do our own duplicate detection here
 
+      if (noForward) {
```

Comment:
If getting to this point the message isn't sent then it would be good to tie a bow on the logging that has occurred to this point and log a trace level message indicating what happened.

Reply:
Also, this seems to be too early to check this. It cant know whether 'mirror forwarding' will occur until during the routing stages later in this method, which is where mirroring handling occurs. Returning here seems like it would mean it drops all messages, even from _this_ broker they were being mirrored to.

Reply:
Thanks @gemmellr for pointing that out. I was having some weird behaviors with messages suddenly disappearing, clearly related to what you are mentioning.